Kernel/Arch/i386/APIC.cpp

```diff
@@ -0,0 +1,128 @@
+{
+    *reinterpret_cast<volatile u32*>(&g_apic_base[off]) = val;
+}
+
+#define APIC_LVT_MASKED (1 << 15)
+#define APIC_LVT_TRIGGER_LEVEL (1 << 14)
+#define APIC_LVT(iv, dm) ((iv & 0xff) | ((dm & 0x7) << 8))
+
+asm(
+    ".globl apic_ap_start \n"
+    ".type apic_ap_start, @function \n"
+    "apic_ap_start: \n"
+    ".set begin_apic_ap_start, . \n"
+    "    jmp apic_ap_start\n" // TODO: implement
+    ".set end_apic_ap_start, . \n"
+    "\n"
+    ".globl apic_ap_start_size \n"
+    "apic_ap_start_size: \n"
+    ".word end_apic_ap_start - begin_apic_ap_start \n");
+
+extern "C" void apic_ap_start(void);
+extern "C" u16 apic_ap_start_size;
+bool init()
+{
+    if (!MSR::have())
+        return false;
+
+    CPUID id(1);
+    if ((id.edx() & (1 << 9)) == 0)
+        return false;
+
+    PhysicalAddress apic_base = get_base();
+    kprintf("Initializing APIC, base: P%x\n", apic_base);
+    set_base(apic_base);
+
+    MM.map_for_kernel(VirtualAddress(apic_base.get()), apic_base, true); // Map memory, disable cache!
+    g_apic_base = apic_base.as_ptr();
+    
+    // copy ap init code to P8000
+    MM.map_for_kernel(VirtualAddress(0x8000), PhysicalAddress(0x8000));
+    memcpy(reinterpret_cast<u8*>(0x8000), reinterpret_cast<const u8*>(apic_ap_start), apic_ap_start_size);
+    return true;
+}
+
+void enable(u32 cpu)
+{
+    kprintf("Enabling local APIC for cpu #%u\n", cpu);
+    
+    // set spurious interrupt vector
+    apic_write(0xf0, apic_read(0xf0) | 0x100);
+    
+    // local destination mode (flat mode)
+    apic_write(0xe0, 0xf000000);
+    
+    // set destination id (note that this limits it to 8 cpus)
+    apic_write(0xd0, (1 << cpu) << 24);
+    
+    register_interrupt_handler(IRQ_APIC_SPURIOUS, apic_spurious_interrupt_entry);
+    
+    apic_write(0x320, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // timer
+    apic_write(0x330, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // thermal
+    apic_write(0x340, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // performance counter
+    apic_write(0x350, APIC_LVT(0x1f, 7) | APIC_LVT_MASKED); // lint0
+    apic_write(0x360, APIC_LVT(0xff, 4) | APIC_LVT_TRIGGER_LEVEL); // lint1 (nmi)
+    apic_write(0x370, APIC_LVT(0xe3, 0) | APIC_LVT_MASKED); // error
+    apic_write(0x500, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // extended int0
+    apic_write(0x510, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // extended int1
+    apic_write(0x520, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // extended int2
+    apic_write(0x530, APIC_LVT(0xff, 0) | APIC_LVT_MASKED); // extended int3
+    
+    if (cpu == 0) {
+        static volatile u32 foo = 0;
+        
+        // INIT
+        apic_write(0x310, 0);
+        apic_write(0x300, 0xc4500); // init all physical (excluding self)
+        for (int i = 0; i < 2; i++) {
+			for (foo = 0; foo < 0x800000; foo++); // TODO: better way of waiting...
+            // SIPI
+            apic_write(0x310, 0);
+            apic_write(0x300, 0xc4608); // start execution at P8000
```

Comment:
There's a huge amount of magic numbers in this patch, can we make this a little bit nicer with some constants/enums?